packages/core/src/common/message-rpc/rpc-protocol.ts

```diff
@@ -0,0 +1,280 @@
+// *****************************************************************************
+// Copyright (C) 2021 Red Hat, Inc. and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { Emitter, Event } from '../event';
+import { Deferred } from '../promise-util';
+import { Channel, MessageProvider } from './channel';
+import { ReadBuffer } from './message-buffer';
+import { RpcMessage, RpcMessageDecoder, RpcMessageEncoder, RpcMessageType } from './rpc-message-encoder';
+import { CancellationToken, CancellationTokenSource } from '../../../shared/vscode-languageserver-protocol';
+
+/**
+ * Handles request messages received by the {@link RpcServer}.
+ */
+export type RequestHandler = (method: string, args: any[]) => Promise<any>;
+
+const CANCELLATION_TOKEN_KEY = 'add.cancellation.token';
+/**
+ * Initialization options for {@link RpcClient}s and {@link RpcServer}s.
+ */
+export interface RpcInitializationOptions extends RPCConnectionOptions {
+    /**
+     * Boolean flag to indicate whether the client/server should be used as as standalone component or is part of
+     * a {@link RpcConnection}. Default is `true`
+     */
+    standalone?: boolean,
+}
+
+/**
+ * A `RpcServer` reads rcp request and notification messages and sends the reply values or
+ * errors from the request to the channel.
+ *  It can either be instantiated as a standalone component or as part of a {@link RpcConnection}.
+ */
+export class RpcServer {
+    protected readonly encoder: RpcMessageEncoder;
+    protected readonly decoder: RpcMessageDecoder;
+
+    protected readonly onNotificationEmitter: Emitter<{ method: string; args: any[]; }> = new Emitter();
+    protected readonly cancellationTokenSources = new Map<number, CancellationTokenSource>();
+
+    get onNotification(): Event<{ method: string; args: any[]; }> {
+        return this.onNotificationEmitter.event;
+    }
+
+    constructor(protected channel: Channel, public readonly requestHandler: RequestHandler, options: RpcInitializationOptions = {}) {
```

Comment:
It might not make sense for the Theia extension Messaging API. However the Plugin ext/main stuff is currently using the separated approach (i.e. `RpcServer` and `RpcClient` instances are created separately). So I guess we should not merge them, unless we also want to completely revamp the plugin messaging approach.